crates/count-min-sketch/src/lib.rs

```diff
@@ -0,0 +1,314 @@
+//! Count-Min Sketch implementation in Rust
+//!
+//! Based on the paper:
+//! <http://dimacs.rutgers.edu/~graham/pubs/papers/cm-full.pdf>
+
+#[allow(deprecated)]
+use std::{
+    fmt,
+    hash::{Hash, Hasher, SipHasher},
+};
+
+use serde::{Deserialize, Serialize};
+
+/// The CountMinHashFn is a data structure used to hash items that are being
+/// added to a Count-Min Sketch.
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
+#[repr(C)]
+pub struct CountMinHashFn {
+    key: u64,
+}
+
+impl CountMinHashFn {
+    /// Creates a new CountMinHashFn whose hash function key is equal to `key`.
+    pub fn with_key(key: u64) -> Self {
+        Self { key }
+    }
+
+    /// Computes the hash of `item` according to the hash function and returns
+    /// the bucket index corresponding to the hashed value.
+    ///
+    /// The returned value will be between 0 and (`nbuckets` - 1).
+    #[allow(deprecated)]
+    pub fn hash_into_buckets<T: Hash>(&self, item: &T, nbuckets: usize) -> usize {
+        let (key1, key2) = (self.key, 0x517cc1b727220a95);
+        let mut hasher = SipHasher::new_with_keys(key1, key2);
+        item.hash(&mut hasher);
+        let hash_val = hasher.finish();
+        (hash_val % (nbuckets as u64)) as usize
+    }
+
+    /// Returns the key for the hash function.
+    pub(crate) fn key(&self) -> u64 {
+        self.key
+    }
+}
+
+/// The Count-Min Sketch is a compact summary data structure capable of
+/// representing a high-dimensional vector and answering queries on this vector,
+/// in particular point queries and dot product queries, with strong accuracy
+/// guarantees. Such queries are at the core of many computations, so the
+/// structure can be used in order to answer a variety of other queries, such as
+/// frequent items (heavy hitters), quantile finding, join size estimation, and
+/// more. Since the data structure can easily process updates in the form of
+/// additions or subtractions to dimensions of the vector (which may correspond
+/// to insertions or deletions, or other transactions), it is capable of working
+/// over streams of updates, at high rates.[1]
+///
+/// [1]: <http://dimacs.rutgers.edu/~graham/pubs/papers/cmencyc.pdf>
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct CountMinSketch {
+    width: usize,
+    depth: usize,
+    hashfuncs: Vec<CountMinHashFn>,
+    counters: Vec<Vec<i64>>,
```

Comment:
Is it worth specifically calling out right here that hashfuncs and counters must both be at least `depth` long? I see both places you construct it now enforce it, and I guess there's little chance we could introduce a third construction that wouldn't explode in testing but... well call me paranoid :)

Reply:
I don't mind enabling a little paranoia :)

Do you want a comment here? in the docstring? 

I spent a little while trying to make something like this work:
```rust
struct CountMinSketch<const D: usize, const W: usize> {
    hashfuncs: [CountMinHashFn; D],
    counters: [[i64; W]; D]
}
```
but I don't think that'd work in the extension.

Reply:
I don't think you can get around run-time assertions. I just like to have documentation too.

In the doc string is good as callers can see it live if their editor is smart enough, and it also allows you to reference other fields in a way that `cargo doc` will check in case we miss the comments on a rename (e.g. `/// must be at least [Self::depth] long`).

Thanks!